courses/advanced-ada/chapters/strong_typing.rst

```diff
@@ -16,6 +16,344 @@ Strong typing
+       end Value;
+
+    end Taint;
+
+Then, make sure the sensitive program uses trusted data:
+
+.. code:: ada
+
+    with Taint; use Taint;
+
+    procedure Sensitive (X : Trusted_Value) is
+    begin
+       null; --  Do something sensitive with value X
+    end Sensitive;
+
+Let's try to pass in data from the user to the sensitive program:
+
+.. code:: ada run_button
+    :class: ada-expect-compile-error
+
+    with Taint;
+    with Sensitive;
+
+    procedure Main is
+
+       procedure Bad (Some_Value : Integer) is
+       begin
+          Sensitive (Some_Value);
+       end Bad;
+
+       A : Integer := 0;
+    begin
+       Bad (A);
+    end Main;
+
+The compiler returns with a type error.
+
+Now, this does not prevent us from doing useful computations on trusted
+data as easily as on tainted data, including initialization with literals,
+case statements, array indexing, etc.
+
+.. code:: ada run_button
+
+    with Taint; use Taint;
+    with Sensitive;
+
+    procedure Good is
+       Max_Value : constant := 100;
+       X : Trusted_Value := Max_Value;
+    begin
+       X := X + 1; --  Perform any computations on X
+       Sensitive (X);
+    end Good;
+
+Because :ada:`Trusted_Value` is a type derived from the tainted type
+(:ada:`Integer`), all operations allowed on tainted data are also allowed
+on trusted data, but operations mixing them are not allowed.
+
+Be aware that nothing prevents the program itself from converting between
+tainted data and trusted data freely, but this requires inserting an
+explicit conversion, which can be spotted during code reviews.
+
+To completely prevent such unintended conversions (say, to facilitate
+maintenance), the type used for trusted data must be made private, so that
+only the package which defines it can convert to and from it. With
+:ada:`Trusted_Value` being private, we should also provide a corresponding
+function for each literal which we used previously, as well as the
+operations that we'd like to allow on trusted values (note that for
+efficiency all operations could be inlined):
+
+.. code:: ada
+
+    package Taint is
+
+       type Trusted_Value is private;
+
+       function Value (V : Trusted_Value) return Integer;
+
+       function Trusted_1 return Trusted_Value;
+       function Trusted_100 return Trusted_Value;
+
+       function "+" (V, W : Trusted_Value) return Trusted_Value;
+
+    private
+
+       type Trusted_Value is new Integer;
+
+    end Taint;
+
+The new implementation is as expected:
+
+.. code:: ada
+
+    package body Taint is
+
+       function Value (V : Trusted_Value) return Integer is
+       begin
+          return Integer (V);
+       end Value;
+
+       function Trusted_1 return Trusted_Value is
+       begin
+          return 1;
+       end Trusted_1;
+
+       function Trusted_100 return Trusted_Value is
+       begin
+          return 100;
+       end Trusted_100;
+
+       function "+" (V, W : Trusted_Value) return Trusted_Value is
```

Comment:
not sure you need to provide an explicit operation here, it should be available already as Trusted_Value is derived from Integer.